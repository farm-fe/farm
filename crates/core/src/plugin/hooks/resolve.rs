use farmfe_macro_cache_item::cache_item;

use crate::module::ModuleId;
use crate::HashMap;

#[derive(Debug, Clone, Hash, PartialEq, Eq, serde::Serialize, serde::Deserialize, Default)]
#[serde(rename_all = "camelCase")]
#[cache_item]
pub enum ResolveKind {
  /// entry input in the config
  Entry(String),
  /// dynamic entry input, which will always be bundled as a separate resource
  /// all deep dependencies of a dynamic entry will be merged into a special module group
  DynamicEntry {
    name: String,
    /// the same as config.output.filename, default to config.output.filename
    output_filename: Option<String>,
  },
  /// static import, e.g. `import a from './a'`
  #[default]
  Import,
  /// static export, e.g. `export * from './a'`
  ExportFrom,
  /// dynamic import, e.g. `import('./a').then(module => console.log(module))`
  DynamicImport,
  /// cjs require, e.g. `require('./a')`
  Require,
  /// @import of css, e.g. @import './a.css'
  CssAtImport,
  /// url() of css, e.g. url('./a.png')
  CssUrl,
  /// `<script src="./index.html" />` of html
  ScriptSrc,
  /// `<link href="index.css" />` of html
  LinkHref,
  /// Hmr update
  HmrUpdate,
  /// Custom ResolveKind
  Custom(String),
}

impl ResolveKind {
  /// dynamic if self is [ResolveKind::DynamicImport] or [ResolveKind::Custom("dynamic:xxx")] (dynamic means the module is loaded dynamically, for example, fetch from network)
  /// used when analyzing module groups
  pub fn is_dynamic_import(&self) -> bool {
    matches!(self, ResolveKind::DynamicImport)
      || matches!(self, ResolveKind::Custom(c) if c.starts_with("dynamic:"))
  }

  pub fn is_dynamic_entry(&self) -> bool {
    matches!(self, ResolveKind::DynamicEntry { .. })
  }

  pub fn is_export_from(&self) -> bool {
    matches!(self, ResolveKind::ExportFrom)
  }

  pub fn is_require(&self) -> bool {
    matches!(self, ResolveKind::Require)
  }
}

impl From<&str> for ResolveKind {
  fn from(value: &str) -> Self {
    serde_json::from_str(value).unwrap()
  }
}

impl From<ResolveKind> for String {
  fn from(value: ResolveKind) -> Self {
    serde_json::to_string(&value).unwrap()
  }
}

/// Parameter of the resolve hook
#[derive(Debug, Clone, serde::Serialize, serde::Deserialize, Hash, PartialEq, Eq)]
#[serde(rename_all = "camelCase")]
pub struct PluginResolveHookParam {
  /// the source would like to resolve, for example, './index'
  pub source: String,
  /// the start location to resolve `specifier`, being [None] if resolving a entry or resolving a hmr update.
  pub importer: Option<ModuleId>,
  /// for example, [ResolveKind::Import] for static import (`import a from './a'`)
  pub kind: ResolveKind,
}

#[derive(Debug, serde::Serialize, serde::Deserialize, Clone)]
#[serde(rename_all = "camelCase", default)]
pub struct PluginResolveHookResult {
  /// resolved path, normally a absolute file path.
  pub resolved_path: String,
  /// whether this module should be external, if true, the module won't present in the final result
  pub external: bool,
  /// whether this module has side effects, affects tree shaking. By default, it's true, means all modules may has side effects.
  /// use sideEffects field in package.json to mark it as side effects free
  pub side_effects: bool,
  /// the query parsed from specifier, for example, query should be `{ inline: "" }` if specifier is `./a.png?inline`
  /// if you custom plugins, your plugin should be responsible for parsing query
  /// if you just want a normal query parsing like the example above, [farmfe_toolkit::resolve::parse_query] should be helpful
  pub query: Vec<(String, String)>,
  /// the meta data passed between plugins and hooks
  pub meta: HashMap<String, String>,
}

impl Default for PluginResolveHookResult {
  fn default() -> Self {
    Self {
      side_effects: true,
      resolved_path: "unknown".to_string(),
      external: false,
      query: vec![],
      meta: Default::default(),
    }
  }
}
