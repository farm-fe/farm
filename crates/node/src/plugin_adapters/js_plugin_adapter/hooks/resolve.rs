use napi::{bindgen_prelude::FromNapiValue, NapiRaw};
use std::sync::Arc;

use farmfe_core::{
  config::config_regex::ConfigRegex,
  context::CompilationContext,
  error::Result,
  plugin::{PluginHookContext, PluginResolveHookParam, PluginResolveHookResult},
};

use crate::{
  new_js_plugin_hook,
  plugin_adapters::js_plugin_adapter::thread_safe_js_plugin_hook::ThreadSafeJsPluginHook,
};

pub struct JsPluginResolveHook {
  tsfn: ThreadSafeJsPluginHook,
  filters: PluginResolveHookFilters,
}

impl JsPluginResolveHook {
  new_js_plugin_hook!(
    PluginResolveHookFilters,
    JsPluginResolveHookFilters,
    PluginResolveHookParam,
    PluginResolveHookResult
  );

  pub fn call(
    &self,
    param: PluginResolveHookParam,
    ctx: Arc<CompilationContext>,
    hook_context: PluginHookContext,
  ) -> Result<Option<PluginResolveHookResult>> {
    let filtered = self.filters.importers.iter().any(|i| {
      if let Some(importer) = &param.importer {
        i.is_match(&importer.resolved_path_with_query(&ctx.config.root))
      } else {
        i.is_match("None")
      }
    }) && self
      .filters
      .sources
      .iter()
      .any(|f| f.is_match(&param.source));

    if filtered {
      self
        .tsfn
        .call::<PluginResolveHookParam, PluginResolveHookResult>(param, ctx, Some(hook_context))
    } else {
      Ok(None)
    }
  }
}

/// Resolve hook filters, works as `||`. If any importers or sources matches any regex item in the Vec, we treat it as filtered.
#[napi(object)]
pub struct JsPluginResolveHookFilters {
  pub importers: Vec<String>,
  pub sources: Vec<String>,
}

#[derive(Debug)]
struct PluginResolveHookFilters {
  pub importers: Vec<ConfigRegex>,
  pub sources: Vec<ConfigRegex>,
}

impl From<JsPluginResolveHookFilters> for PluginResolveHookFilters {
  fn from(f: JsPluginResolveHookFilters) -> Self {
    Self {
      importers: f
        .importers
        .into_iter()
        .map(|f| ConfigRegex::new(&f))
        .collect(),
      sources: f
        .sources
        .into_iter()
        .map(|f| ConfigRegex::new(&f))
        .collect(),
    }
  }
}
