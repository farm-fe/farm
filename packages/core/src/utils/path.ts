import path from 'node:path';

const windowsSlashRE = /\\/g;
export function slash(p: string): string {
  return p.replace(windowsSlashRE, '/');
}

export function withTrailingSlash(path: string): string {
  if (path[path.length - 1] !== '/') {
    return `${path}/`;
  }
  return path;
}

const postfixRE = /[?#].*$/;
export function stripQueryAndHash(path: string): string {
  return path.replace(postfixRE, '');
}

export function removeHashFromPath(path: string): string {
  const hashRE = /([_-][a-f0-9]{4,12})(\.[^./]+(\.[^./]+)*)$/;
  return path.replace(hashRE, '$2');
}

export function getShortName(file: string, root: string): string {
  const result = file.startsWith(withTrailingSlash(root))
    ? path.relative(root, file)
    : file;
  return result;
}
