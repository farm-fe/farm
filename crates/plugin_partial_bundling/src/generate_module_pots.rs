use std::path::PathBuf;

use farmfe_core::{
  config::partial_bundling::PartialBundlingConfig,
  hashbrown::{HashMap, HashSet},
  module::{module_graph::ModuleGraph, Module, ModuleId},
};
use farmfe_toolkit::resolve::load_package_json;

use crate::{generate_module_buckets::ResourceType, module_pot::ModulePot};

pub fn generate_module_pots(
  modules: &HashSet<ModuleId>,
  module_graph: &ModuleGraph,
  config: &PartialBundlingConfig,
  root: &str,
  resource_type: ResourceType,
) -> Vec<ModulePot> {
  let mut module_pot_map = HashMap::<String, ModulePot>::new();

  for module_id in modules {
    let module = module_graph.module(module_id).unwrap();
    let module_pot_name = generate_module_pot_name(module, config, root, resource_type.clone());
    let module_pot_id = ModulePot::gen_id(
      &module_pot_name,
      module.module_type.clone(),
      module.immutable,
    );

    let module_pot = module_pot_map.entry(module_pot_id).or_insert_with(|| {
      ModulePot::new(
        module_pot_name,
        module.module_type.clone(),
        module.immutable,
      )
    });

    module_pot.add_module(module_id.clone(), module.size, module.execution_order);
  }

  // split module_pots from module_pot_map that its size larger that target_max_size
  let exceed_size_module_pot_ids = module_pot_map
    .iter()
    .filter(|(_, module_pot)| module_pot.size > config.target_max_size)
    .map(|(module_pot_id, _)| module_pot_id.clone())
    .collect::<HashSet<_>>();

  for exceed_size_module_pot_id in exceed_size_module_pot_ids {
    let module_pot = module_pot_map.remove(&exceed_size_module_pot_id).unwrap();
    let new_module_pot_numbers = (module_pot.size / config.target_max_size) + 1;
    let module_pot_name = module_pot.name.clone();
    let immutable = module_pot.immutable;
    let ty = module_pot.module_type.clone();
    let mut modules = module_pot.take_modules().into_iter().collect::<Vec<_>>();
    modules.sort_by_key(|m| m.to_string());
    let page_size = modules.len() / new_module_pot_numbers;

    for i in 0..new_module_pot_numbers {
      let new_module_pot_name = format!("{}-{}", module_pot_name, i);
      let new_module_pot_id = ModulePot::gen_id(&new_module_pot_name, ty.clone(), immutable);

      let new_module_pot = module_pot_map
        .entry(new_module_pot_id)
        .or_insert_with(|| ModulePot::new(new_module_pot_name, ty.clone(), immutable));

      let start = i * page_size;
      let end = if i == new_module_pot_numbers - 1 {
        modules.len()
      } else {
        (i + 1) * page_size
      };

      for module_id in &modules[start..end] {
        let module = module_graph.module(module_id).unwrap();
        new_module_pot.add_module(module_id.clone(), module.size, module.execution_order);
      }
    }
  }

  let mut module_pots = module_pot_map
    .into_iter()
    .map(|(_, module_pot)| module_pot)
    .collect::<Vec<_>>();

  module_pots.sort_by_key(|m| m.execution_order);

  module_pots
}

fn generate_module_pot_name(
  module: &Module,
  config: &PartialBundlingConfig,
  root: &str,
  resource_type: ResourceType,
) -> String {
  // 1. get name from partialBundling.groups
  for group_config in &config.groups {
    // use the first matched group name, so the order of groups is important
    if group_config
      .test
      .iter()
      .any(|c| c.is_match(&module.id.to_string()))
    {
      if (group_config.group_type.is_match(module.immutable))
        && (resource_type.is_match(group_config.resource_type.clone()))
      {
        return group_config.name.clone();
      }
    }
  }

  // 2. get name from immutable package
  if module.immutable {
    let package_json = load_package_json(
      PathBuf::from(module.id.resolved_path(root)),
      Default::default(),
    );

    if let Ok(package_json) = package_json {
      let mut name = package_json.name.unwrap_or_default();

      if let Some(version) = package_json.version {
        name = format!("{}@{}", name, version);
      }

      return name;
    }
  }

  module.id.to_string()
}

#[cfg(test)]
mod tests {
  use farmfe_core::{
    config::{
      config_regex::ConfigRegex,
      partial_bundling::{
        PartialBundlingConfig, PartialBundlingGroupConfig, PartialBundlingGroupConfigGroupType,
        PartialBundlingGroupConfigResourceType,
      },
    },
    hashbrown::HashSet,
    module::{module_graph::ModuleGraph, Module, ModuleType},
  };
  use farmfe_testing_helpers::fixture;

  use crate::{
    generate_module_buckets::ResourceType, generate_module_pots::generate_module_pots,
    module_pot::ModulePot,
  };

  #[test]
  fn test_generate_module_pots_package() {
    fixture!(
      "tests/fixtures/generate_module_pots/basic/index.ts",
      |_, cwd| {
        let mut module_graph = ModuleGraph::new();
        let mut module_index =
          Module::new("tests/fixtures/generate_module_pots/basic/index.ts".into());
        let mut module_utils =
          Module::new("tests/fixtures/generate_module_pots/basic/utils.ts".into());
        let mut module_a =
          Module::new("tests/fixtures/generate_module_pots/basic1/index.ts".into());

        module_index.module_type = ModuleType::Js;
        module_index.immutable = true;
        module_index.size = 10 * 1024;
        module_index.execution_order = 2;

        module_utils.module_type = ModuleType::Js;
        module_utils.immutable = true;
        module_utils.size = 5 * 1024;
        module_utils.execution_order = 3;

        module_a.module_type = ModuleType::Js;
        module_a.execution_order = 1;
        module_a.immutable = true;
        module_a.size = 1 * 1024;

        module_graph.add_module(module_index);
        module_graph.add_module(module_utils);
        module_graph.add_module(module_a);

        let modules = module_graph
          .modules()
          .iter()
          .map(|m| m.id.clone())
          .collect::<HashSet<_>>();

        let module_pots = generate_module_pots(
          &modules,
          &module_graph,
          &Default::default(),
          cwd.to_str().unwrap(),
          ResourceType::Initial,
        );

        assert_eq!(module_pots.len(), 2);
        assert_eq!(module_pots[0].name, "test-package1@1.0.0");
        assert_eq!(module_pots[0].size, 1 * 1024);
        assert_eq!(module_pots[0].module_type, ModuleType::Js);
        assert_eq!(module_pots[0].immutable, true);
        assert_eq!(module_pots[0].execution_order, 1);
        assert_eq!(
          module_pots[0].modules(),
          &HashSet::from(["tests/fixtures/generate_module_pots/basic1/index.ts".into()])
        );

        assert_eq!(module_pots[1].name, "test-package@1.0.0");
        assert_eq!(module_pots[1].size, 15 * 1024);
        assert_eq!(module_pots[1].module_type, ModuleType::Js);
        assert_eq!(module_pots[1].immutable, true);
        assert_eq!(module_pots[1].execution_order, 2);
        assert_eq!(
          module_pots[1].modules(),
          &HashSet::from([
            "tests/fixtures/generate_module_pots/basic/index.ts".into(),
            "tests/fixtures/generate_module_pots/basic/utils.ts".into()
          ])
        );
      }
    );
  }

  #[test]
  fn test_generate_module_pots_group() {
    let mut module_graph = ModuleGraph::new();
    let mut module_a = Module::new("src/a.ts".into());
    let mut module_b = Module::new("src/b.ts".into());
    let mut module_c = Module::new("utils/c.ts".into());

    module_a.module_type = ModuleType::Js;
    module_a.size = 10 * 1024;
    module_a.execution_order = 1;

    module_b.module_type = ModuleType::Js;
    module_b.size = 5 * 1024;
    module_b.execution_order = 2;

    module_c.module_type = ModuleType::Js;
    module_c.size = 1 * 1024;
    module_c.execution_order = 3;

    module_graph.add_module(module_a);
    module_graph.add_module(module_b);
    module_graph.add_module(module_c);

    let modules = module_graph
      .modules()
      .iter()
      .map(|m| m.id.clone())
      .collect::<HashSet<_>>();

    let assert_group_works = |module_pots: Vec<ModulePot>| {
      assert_eq!(module_pots.len(), 2);

      assert_eq!(module_pots[0].name, "test");
      assert_eq!(module_pots[0].size, 15 * 1024);
      assert_eq!(module_pots[0].module_type, ModuleType::Js);
      assert_eq!(module_pots[0].immutable, false);
      assert_eq!(module_pots[0].execution_order, 1);
      assert_eq!(
        module_pots[0].modules(),
        &HashSet::from(["src/a.ts".into(), "src/b.ts".into()])
      );

      assert_eq!(module_pots[1].name, "utils/c.ts");
      assert_eq!(module_pots[1].size, 1 * 1024);
      assert_eq!(module_pots[1].module_type, ModuleType::Js);
      assert_eq!(module_pots[1].immutable, false);
      assert_eq!(module_pots[1].execution_order, 3);
      assert_eq!(
        module_pots[1].modules(),
        &HashSet::from(["utils/c.ts".into()])
      );
    };

    let assert_group_not_works = |module_pots: Vec<ModulePot>| {
      assert_eq!(module_pots.len(), 3);

      assert_eq!(module_pots[0].name, "src/a.ts");
      assert_eq!(module_pots[0].size, 10 * 1024);
      assert_eq!(module_pots[0].module_type, ModuleType::Js);
      assert_eq!(module_pots[0].immutable, false);
      assert_eq!(module_pots[0].execution_order, 1);
      assert_eq!(
        module_pots[0].modules(),
        &HashSet::from(["src/a.ts".into()])
      );

      assert_eq!(module_pots[1].name, "src/b.ts");
      assert_eq!(module_pots[1].size, 5 * 1024);
      assert_eq!(module_pots[1].module_type, ModuleType::Js);
      assert_eq!(module_pots[1].immutable, false);
      assert_eq!(module_pots[1].execution_order, 2);
      assert_eq!(
        module_pots[1].modules(),
        &HashSet::from(["src/b.ts".into()])
      );

      assert_eq!(module_pots[2].name, "utils/c.ts");
      assert_eq!(module_pots[2].size, 1 * 1024);
      assert_eq!(module_pots[2].module_type, ModuleType::Js);
      assert_eq!(module_pots[2].immutable, false);
      assert_eq!(module_pots[2].execution_order, 3);
      assert_eq!(
        module_pots[2].modules(),
        &HashSet::from(["utils/c.ts".into()])
      );
    };

    // Default config for group_type and resource_type
    let config = PartialBundlingConfig {
      groups: vec![PartialBundlingGroupConfig {
        name: "test".into(),
        test: vec![ConfigRegex::new("src/.*")],
        ..Default::default()
      }],
      ..Default::default()
    };

    let module_pots = generate_module_pots(
      &modules,
      &module_graph,
      &config,
      "/root",
      ResourceType::Initial,
    );
    assert_group_works(module_pots);

    // only match mutable modules
    let config = PartialBundlingConfig {
      groups: vec![PartialBundlingGroupConfig {
        name: "test".into(),
        test: vec![ConfigRegex::new("src/.*")],
        group_type: PartialBundlingGroupConfigGroupType::Immutable,
        ..Default::default()
      }],
      ..Default::default()
    };

    let module_pots = generate_module_pots(
      &modules,
      &module_graph,
      &config,
      "/root",
      ResourceType::Initial,
    );

    assert_group_not_works(module_pots);

    let config = PartialBundlingConfig {
      groups: vec![PartialBundlingGroupConfig {
        name: "test".into(),
        test: vec![ConfigRegex::new("src/.*")],
        resource_type: PartialBundlingGroupConfigResourceType::Async,
        ..Default::default()
      }],
      ..Default::default()
    };

    let module_pots = generate_module_pots(
      &modules,
      &module_graph,
      &config,
      "/root",
      ResourceType::Initial,
    );

    assert_group_not_works(module_pots);

    let module_pots = generate_module_pots(
      &modules,
      &module_graph,
      &config,
      "/root",
      ResourceType::Async,
    );

    assert_group_works(module_pots);

    let config = PartialBundlingConfig {
      groups: vec![PartialBundlingGroupConfig {
        name: "test".into(),
        test: vec![ConfigRegex::new("src/.*")],
        resource_type: PartialBundlingGroupConfigResourceType::Initial,
        ..Default::default()
      }],
      ..Default::default()
    };

    let module_pots = generate_module_pots(
      &modules,
      &module_graph,
      &config,
      "/root",
      ResourceType::Initial,
    );

    assert_group_works(module_pots);
  }
}
