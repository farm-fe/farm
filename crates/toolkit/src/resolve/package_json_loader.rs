use std::path::PathBuf;

use farmfe_core::{
  common::PackageJsonInfo,
  dashmap::DashMap,
  error::{CompilationError, Result},
  serde_json::from_str,
};

use crate::fs::read_file_utf8;

use super::follow_symlinks;

const PACKAGE_JSON_FILE: &str = "package.json";

/// Load closest package.json, return [farmfe_core::error::Result] if not found.
/// With cache supported, if the giving path is loaded then the cache will be used.
///
/// ```ignore
/// let package_json_loader = PackageJsonLoader::new();
/// let info = package_json.load("/root/packages/app/src")?;
/// ```
pub struct PackageJsonLoader {
  /// path -> package_json_info cache
  cache: DashMap<String, PackageJsonInfo>,
}

pub struct Options {
  // whether follow symlinks when resolving package.json
  pub follow_symlinks: bool,
  // whether resolve package.json in ancestor directories
  pub resolve_ancestor_dir: bool,
}

impl Default for Options {
  fn default() -> Self {
    Self {
      follow_symlinks: true,
      resolve_ancestor_dir: true,
    }
  }
}

impl PackageJsonLoader {
  pub fn new() -> Self {
    Self {
      cache: DashMap::new(),
    }
  }

  /// resolve package.json start from path to all its ancestor
  pub fn load(&self, path: PathBuf, options: Options) -> Result<PackageJsonInfo> {
    let mut current = path.clone();
    let mut visited_stack = vec![];

    while current.parent().is_some() {
      if self
        .cache
        .contains_key(&current.to_string_lossy().to_string())
      {
        return Ok(
          self
            .cache
            .get(&current.to_string_lossy().to_string())
            .unwrap()
            .clone(),
        );
      }

      // TODO cover it with tests
      if !options.resolve_ancestor_dir && visited_stack.len() == 1 {
        return Err(CompilationError::LoadPackageJsonError {
          package_json_path: current.to_string_lossy().to_string(),
          err_message: "can not load node_modules when resolve_ancestor_dir set to false"
            .to_string(),
        });
      }

      visited_stack.push(current.clone());

      let package_json_path = if options.follow_symlinks {
        follow_symlinks(current.join(PACKAGE_JSON_FILE))
      } else {
        current.join(PACKAGE_JSON_FILE)
      };

      if package_json_path.exists() && package_json_path.is_file() {
        let content = read_file_utf8(package_json_path.to_str().unwrap())?;

        let mut result: PackageJsonInfo =
          from_str(&content).map_err(|e| CompilationError::LoadPackageJsonError {
            package_json_path: package_json_path.to_string_lossy().to_string(),
            err_message: format!("{:?}", e),
          })?;
        result.set_raw_map(from_str(&content).map_err(|e| {
          CompilationError::LoadPackageJsonError {
            package_json_path: package_json_path.to_string_lossy().to_string(),
            err_message: format!("{:?}", e),
          }
        })?);
        result.set_raw(content);
        result.set_dir(
          package_json_path
            .parent()
            .unwrap()
            .to_string_lossy()
            .to_string(),
        );
        result.parse();

        for visited in visited_stack {
          self
            .cache
            .insert(visited.to_string_lossy().to_string(), result.clone());
        }

        return Ok(result);
      }

      current = current.parent().unwrap().to_path_buf();
    }

    Err(CompilationError::LoadPackageJsonError {
      package_json_path: path.to_string_lossy().to_string(),
      err_message: String::from("Can not find package.json in all ancestor directories"),
    })
  }

  pub fn cache(&self) -> &DashMap<String, PackageJsonInfo> {
    &self.cache
  }
}
