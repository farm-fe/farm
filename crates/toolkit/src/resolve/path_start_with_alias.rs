use std::collections::HashMap;

use farmfe_core::regex::Regex;

const REGEX_PREFIX: &str = "$__farm_regex:";

/// Determine whether the path conforms to the configuration of alias.
pub fn is_start_with_alias(alias_map: &HashMap<String, String>, path: &str) -> bool {
  let mut aliases: Vec<&str> = alias_map.keys().map(|k| k.as_str()).collect();
  aliases.sort_by(|a, b| b.len().cmp(&a.len()));

  aliases.iter().any(|&alias| {
    if let Some(stripped_alias) = alias.strip_prefix(REGEX_PREFIX) {
      let regex = Regex::new(stripped_alias).unwrap();
      return regex.is_match(path);
    } else if alias.ends_with("$") && path == alias.trim_end_matches('$') {
      return true;
    } else {
      return path.starts_with(alias);
    }
  })
}

#[cfg(test)]

mod test {
  use std::path::PathBuf;

  use super::*;

  #[test]
  fn test() {
    let cwd = PathBuf::from("/root/src");

    let alias = HashMap::from([
      ("/@".to_string(), cwd.to_string_lossy().to_string()),
      ("@".to_string(), cwd.to_string_lossy().to_string()),
      ("react$".to_string(), cwd.to_string_lossy().to_string()),
      (
        "$__farm_regex:^/(utils)$".to_string(),
        cwd.join("$1").to_string_lossy().to_string(),
      ),
    ]);

    assert_eq!(is_start_with_alias(&alias, "/@/img/logo.png"), true);
    assert_eq!(is_start_with_alias(&alias, "@/img/logo.png"), true);

    assert_eq!(is_start_with_alias(&alias, "./img/logo.png"), false);
    assert_eq!(is_start_with_alias(&alias, "../img/logo.png"), false);
    assert_eq!(is_start_with_alias(&alias, "/img/logo.png"), false);

    assert_eq!(is_start_with_alias(&alias, "react/useEffect.ts"), false);
    assert_eq!(is_start_with_alias(&alias, "react"), true);

    assert_eq!(is_start_with_alias(&alias, "/utils"), true);
    assert_eq!(is_start_with_alias(&alias, "utils"), false);
    assert_eq!(is_start_with_alias(&alias, "/utils/index.ts"), false);
  }
}
