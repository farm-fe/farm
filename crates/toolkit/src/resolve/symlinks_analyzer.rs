use std::path::PathBuf;

use farmfe_core::{dashmap::DashMap, relative_path::RelativePath};
use farmfe_utils::diff_paths;

/// Analyze symlinks and get the real path.
/// It will traverse all the ancestor of the specified path, if any ancestor is symlinks if will be followed to the real path.
/// If it is not symlinked, just return the original path.
pub struct SymlinksAnalyzer {
  cache: DashMap<PathBuf, PathBuf>,
}

impl SymlinksAnalyzer {
  pub fn new() -> Self {
    Self {
      cache: DashMap::new(),
    }
  }

  pub fn follow_symlinks(&self, path: PathBuf) -> PathBuf {
    if self.cache.contains_key(&path) {
      return self.cache.get(&path).unwrap().clone();
    }

    let mut current = path.clone();
    let mut real_path = path.clone();

    while current.parent().is_some() {
      if current.is_symlink() {
        if cfg!(windows) {
          real_path = current.read_link().unwrap();
        } else {
          real_path = RelativePath::new(current.read_link().unwrap().to_str().unwrap())
            .to_logical_path(current.parent().unwrap());
        }

        break;
      }

      current = current.parent().unwrap().to_path_buf();
    }

    // there is symlink existed
    let real_path = if real_path != path {
      let relative_path = diff_paths(path.clone(), current).unwrap();
      real_path.join(relative_path)
    } else {
      path.clone()
    };

    self.cache.insert(path, real_path.clone());

    real_path
  }
}
