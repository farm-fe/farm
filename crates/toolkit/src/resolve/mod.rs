//! Shared utilities for plugins.
//! ```ignore
//! use farmfe_core::load_package_json;
//!
//! struct MyPlugin {}
//!
//! impl Plugin for MyPlugin {
//!   fn resolve(...) -> ... {
//!     let pkg_json = load_package_json(dir);
//!   }
//! }
//! ```

use std::path::PathBuf;

use farmfe_core::common::PackageJsonInfo;
use lazy_static::lazy_static;

pub mod package_json_loader;
pub mod symlinks_analyzer;

use package_json_loader::PackageJsonLoader;

use crate::resolve::symlinks_analyzer::SymlinksAnalyzer;

use self::package_json_loader::Options;

lazy_static! {
  pub static ref PACKAGE_JSON_LOADER: PackageJsonLoader = PackageJsonLoader::new();
  pub static ref SYMLINKS_ANALYZER: SymlinksAnalyzer = SymlinksAnalyzer::new();
}

pub const DYNAMIC_EXTENSION_PRIORITY: &str = "DYNAMIC_EXTENSION_PRIORITY";

/// Load closest package.json start from the specified path, return [farmfe_core::error::Result<Value>].
pub fn load_package_json(
  path: PathBuf,
  options: Options,
) -> farmfe_core::error::Result<PackageJsonInfo> {
  // using global static package.json loader
  PACKAGE_JSON_LOADER.load(path, options)
}

/// The default package.json info is:
/// ```json
/// {
///   "name": "farm-default-package-info",
///   "version": "0.0.0"
/// }
/// ```
/// And this function won't trigger any file/io operation
pub fn default_package_json() -> PackageJsonInfo {
  PackageJsonInfo::new(
    Some("farm-default-package-info".to_string()),
    Some("0.0.0".to_string()),
  )
}

/// Try follow symlinks from the specified path, if any ancestor of the path is symlinked, it will be redirected to the real path.
/// For example, the path is `/root/react/index.js` while `/root/react` is symlinked to `/root/store/react`, then the result should be `/root/store/react/index.js`.
pub fn follow_symlinks(path: PathBuf) -> PathBuf {
  SYMLINKS_ANALYZER.follow_symlinks(path)
}
