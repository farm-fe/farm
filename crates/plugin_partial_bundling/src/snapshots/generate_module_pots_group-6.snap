---
source: crates/plugin_partial_bundling/src/generate_module_pots.rs
input_file: crates/testing_helpers/src/lib.rs
---
(
    ModulePot {
        name: "src/a.ts",
        source_type: MutableModule,
        size: 10240,
        module_type: Js,
        immutable: false,
        execution_order: 1,
        modules: {},
    },
    [
        ModuleId {
            relative_path: "src/a.ts",
            query_string: "",
        },
    ],
)
